File: backend/app/services/site_service.py
from fastapi import HTTPException, status
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.models.site import Site
from app.models.site_species import SiteSpecies
from app.schemas.site import SiteCreate, SiteUpdate


def create_site(db: Session, data: SiteCreate, user_id: int) -> Site:
    site = Site(
        name=data.name,
        code=data.code,
        description=data.description,
        created_by=user_id,
    )
    db.add(site)
    db.commit()
    db.refresh(site)
    return site


def get_site(db: Session, site_id: int) -> Site:
    site = db.query(Site).filter(Site.id == site_id).first()
    if not site:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Site not found")
    return site


def list_sites(db: Session, search: str | None = None) -> list[Site]:
    query = db.query(Site)
    if search:
        query = query.filter(Site.name.ilike(f"%{search}%"))
    return query.order_by(Site.name).all()


def update_site(db: Session, site_id: int, data: SiteUpdate) -> Site:
    site = get_site(db, site_id)
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(site, field, value)
    db.commit()
    db.refresh(site)
    return site


def delete_site(db: Session, site_id: int, force: bool = False) -> None:
    site = get_site(db, site_id)
    has_associations = db.query(SiteSpecies).filter(SiteSpecies.site_id == site_id).first()
    if has_associations and not force:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Site has species associations. Use force=true to delete anyway.",
        )
    db.delete(site)
    db.commit()

File: frontend/src/features/sites/services/siteService.ts
import { api } from '@/services/api'
import type { Site, SiteCreateInput, SiteUpdateInput } from '../types'

export async function listSites(search?: string): Promise<Site[]> {
  const { data } = await api.get<Site[]>('/sites', { params: search ? { search } : undefined })
  return data
}

export async function getSite(siteId: number): Promise<Site> {
  const { data } = await api.get<Site>(`/sites/${siteId}`)
  return data
}

export async function createSite(payload: SiteCreateInput): Promise<Site> {
  const { data } = await api.post<Site>('/sites', payload)
  return data
}

export async function updateSite(siteId: number, payload: SiteUpdateInput): Promise<Site> {
  const { data } = await api.patch<Site>(`/sites/${siteId}`, payload)
  return data
}

export async function deleteSite(siteId: number, force = false): Promise<void> {
  await api.delete(`/sites/${siteId}`, { params: force ? { force: true } : undefined })
}
